server/sonar-server/src/main/java/org/sonar/server/computation/sqale/SqaleNewMeasuresVisitor.java

```diff
@@ -0,0 +1,249 @@
+  private final PeriodsHolder periodsHolder;
+  private final SqaleRatingSettings sqaleRatingSettings;
+
+  private final Metric newDebtMetric;
+  private final Metric nclocDataMetric;
+  private final Metric newDebtRatioMetric;
+
+  public SqaleNewMeasuresVisitor(MetricRepository metricRepository, MeasureRepository measureRepository, BatchReportReader batchReportReader, PeriodsHolder periodsHolder,
+    SqaleRatingSettings sqaleRatingSettings) {
+    super(CrawlerDepthLimit.FILE, POST_ORDER, NewDevelopmentCostCounterFactory.INSTANCE);
+    this.batchReportReader = batchReportReader;
+    this.measureRepository = measureRepository;
+    this.periodsHolder = periodsHolder;
+    this.sqaleRatingSettings = sqaleRatingSettings;
+
+    // computed by NewDebtAggregator which is executed by IntegrateIssuesVisitor
+    this.newDebtMetric = metricRepository.getByKey(CoreMetrics.NEW_TECHNICAL_DEBT_KEY);
+    // which line is ncloc and which isn't
+    this.nclocDataMetric = metricRepository.getByKey(CoreMetrics.NCLOC_DATA_KEY);
+    // output metric
+    this.newDebtRatioMetric = metricRepository.getByKey(CoreMetrics.NEW_SQALE_DEBT_RATIO_KEY);
+  }
+
+  @Override
+  public void visitProject(Component project, Path<NewDevelopmentCostCounter> path) {
+    computeAndSaveNewDebtRatioMeasure(project, path);
+  }
+
+  @Override
+  public void visitModule(Component module, Path<NewDevelopmentCostCounter> path) {
+    computeAndSaveNewDebtRatioMeasure(module, path);
+    increaseNewDevCostOfParent(path);
+  }
+
+  @Override
+  public void visitDirectory(Component directory, Path<NewDevelopmentCostCounter> path) {
+    computeAndSaveNewDebtRatioMeasure(directory, path);
+    increaseNewDevCostOfParent(path);
+  }
+
+  @Override
+  public void visitFile(Component file, Path<NewDevelopmentCostCounter> path) {
+    if (file.getFileAttributes().isUnitTest()) {
+      return;
+    }
+
+    initNewDebtRatioCounter(file, path);
+    computeAndSaveNewDebtRatioMeasure(file, path);
+    increaseNewDevCostOfParent(path);
+  }
+
+  private void computeAndSaveNewDebtRatioMeasure(Component component, Path<NewDevelopmentCostCounter> path) {
+    MeasureVariations.Builder builder = newMeasureVariationsBuilder();
+    for (Period period : periodsHolder.getPeriods()) {
+      long newDevCost = path.current().getValue(period).getValue();
+      double density = computeDensity(component, period, newDevCost);
+      builder.setVariation(period, 100.0 * density);
+    }
+    if (!builder.isEmpty()) {
+      Measure measure = newMeasureBuilder().setVariations(builder.build()).createNoValue();
+      measureRepository.add(component, this.newDebtRatioMetric, measure);
+    }
+  }
+
+  private double computeDensity(Component component, Period period, long developmentCost) {
+    double debt = getLongValue(measureRepository.getRawMeasure(component, this.newDebtMetric), period);
+    if (developmentCost != 0L) {
+      return debt / (double) developmentCost;
+    }
+    return 0d;
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Upper-case this literal "d" suffix. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS818)